app/controllers/images_controller.rb

```diff
@@ -0,0 +1,328 @@
+#  reuse_image::           Choose images to add to observation.
+#  remove_images::         Choose images to remove from observation.
+#  remove_images_for_glossary_term Choose images to remove from GlossaryTerm.
+#  license_updater::       Change copyright of many images.
+#  bulk_vote_anonymity_updater:: Change anonymity of image votes in bulk.
+#  bulk_filename_purge::   Purge all original image filenames from the database.
+#  process_image::         (helper for add_image)
+#
+class ImagesController < ApplicationController
+  before_action :login_required
+  before_action :pass_query_params, except: [:index]
+  before_action :disable_link_prefetching, except: [:show]
+
+  ##############################################################################
+  #
+  #  :section: Searches and Indexes
+  #
+  ##############################################################################
+
+  def index # rubocop:disable Metrics/AbcSize
+    if params[:advanced_search].present?
+      advanced_search
+    elsif params[:pattern].present?
+      image_search
+    elsif params[:by_user].present?
+      images_by_user
+    elsif params[:for_project].present?
+      images_for_project
+    elsif params[:by].present?
+      index_image
+    else
+      list_images
+    end
+  end
+
+  private
+
+  # Display matrix of selected images, based on current Query.
+  def index_image
+    query = find_or_create_query(:Image, by: params[:by])
+    show_selected_images(query, id: params[:id].to_s, always_index: true)
+  end
+
+  # Display matrix of images, most recent first.
+  def list_images
+    if params[:page].to_s.to_i > 1000
+      render(
+        status: :too_many_requests,
+        content_type: "text/plain",
+        plain: "Your queries are killing our server. " \
+               "There are much better ways to scrape the images " \
+               "from our site. " \
+               "Please contact the webmaster." \
+               "And please stop hammering our server!"
+      )
+      return
+    end
+
+    query = create_query(:Image, :all, by: :created_at)
+    show_selected_images(query)
+  end
+
+  # Display matrix of images by a given user.
+  def images_by_user
+    user = if params[:by_user]
+             find_or_goto_index(User, params[:by_user].to_s)
+           else
+             @user
+           end
+    return unless user
+
+    query = create_query(:Image, :by_user, user: user)
+    show_selected_images(query)
+  end
+
+  # Display matrix of Image's attached to a given project.
+  def images_for_project
+    project = find_or_goto_index(Project, params[:for_project].to_s)
+    return unless project
+
+    query = create_query(:Image, :for_project, project: project)
+    show_selected_images(query, always_index: 1)
+  end
+
+  # Display matrix of images whose notes, names, etc. match a string pattern.
+  def image_search
+    pattern = params[:pattern].to_s
+    if pattern.match(/^\d+$/) &&
```

Comment:
```suggestion
    if pattern.match?(/^\d+$/) &&
```
See https://docs.rubocop.org/rubocop-performance/cops_performance.html#performanceregexpmatch
(I don't know why rubocop doesn't flag this.)

Reply:
Oh wow. Great catch. 

These are all probably over the CRUD controllers, in every index. 
I will commit this now, and make another PR to get them elsewhere.

Reply:
Our `.rubocop.yml` may explain why rubocop doesn't catch this:
`something.match?(regexp)` throws an error if `something` is `nil`.
That won't be true here.

Reply:
Can we be sure the pattern won't be `nil`?

Reply:
Yes, the `.to_s` above ensures it is not nil.

Reply:
Thanks @pellaea , that's good to know.